File: backend/RAG/src/vector_search.py
import os
from azure.search.documents import SearchClient
from azure.search.documents.indexes import SearchIndexClient
from azure.search.documents.indexes.models import (
    SearchIndex, SearchField, SearchFieldDataType, SimpleField, SearchableField,
    VectorSearch, HnswAlgorithmConfiguration, VectorSearchProfile, SemanticConfiguration,
    SemanticField, SemanticPrioritizedFields, SemanticSearch
)
from azure.core.credentials import AzureKeyCredential
from dotenv import load_dotenv

load_dotenv()


class VectorSearchManager:
    def __init__(self):
        self.search_service_name = os.getenv("AZURE_SEARCH_SERVICE_NAME")
        self.search_api_key = os.getenv("AZURE_SEARCH_API_KEY")
        self.search_index_name = os.getenv("AZURE_SEARCH_INDEX_NAME")
        self.endpoint = f"https://{self.search_service_name}.search.windows.net"
        self.search_client = None
        self.index_client = None

    def initialize(self):
        """Initialize search client and create index if it doesn't exist."""
        try:
            credential = AzureKeyCredential(self.search_api_key)
            self.search_client = SearchClient(
                endpoint=self.endpoint,
                index_name=self.search_index_name,
                credential=credential
            )
            self.index_client = SearchIndexClient(
                endpoint=self.endpoint,
                credential=credential
            )
            
            # Check if index exists, create if not
            if self._index_exists():
                print(f"Search index '{self.search_index_name}' already exists.")
            else:
                print(f"Search index '{self.search_index_name}' does not exist. Creating...")
                if self._create_index():
                    print(f"Search index '{self.search_index_name}' created successfully.")
                else:
                    print(f"Failed to create index '{self.search_index_name}'.")
                    return False
            return True
        except Exception as e:
            print(f"Error initializing search client: {e}")
            return False

    def _index_exists(self):
        """Check if index exists."""
        try:
            self.index_client.get_index(self.search_index_name)
            return True
        except:
            return False

    def _create_index(self):
        """Create search index with vector field."""
        try:
            index = SearchIndex(
                name=self.search_index_name,
                fields=[
                    SimpleField(name="id", type=SearchFieldDataType.String, key=True),
                    SearchableField(name="content", type=SearchFieldDataType.String, analyzer_name="standard.lucene"),
                    SearchableField(name="source", type=SearchFieldDataType.String, retrievable=True, filterable=True),
                    SearchableField(name="title", type=SearchFieldDataType.String, retrievable=True),
                    SearchField(name="embedding", type=SearchFieldDataType.Collection(SearchFieldDataType.Single),
                               searchable=True, vector_search_dimensions=1536, vector_search_profile_name="myHnswProfile"),
                ],
                vector_search=VectorSearch(
                    algorithms=[HnswAlgorithmConfiguration(name="myHnsw")],
                    profiles=[VectorSearchProfile(name="myHnswProfile", algorithm_configuration_name="myHnsw")]
                ),
                semantic_search=SemanticSearch(
                    configurations=[SemanticConfiguration(
                        name="default",
                        prioritized_fields=SemanticPrioritizedFields(
                            content_fields=[SemanticField(field_name="content")]
                        )
                    )]
                )
            )
            self.index_client.create_index(index)
            return True
        except Exception as e:
            print(f"Error creating index: {e}")
            return False

    def vector_search(self, embedding, top_k=5):
        """Perform vector search on indexed documents."""
        try:
            from azure.search.documents.models import VectorizedQuery
            
            vector_query = VectorizedQuery(vector=embedding, k_nearest_neighbors=top_k, fields="embedding")
            results = self.search_client.search(
                search_text="",
                vector_queries=[vector_query],
                top=top_k,
                include_total_count=True
            )
            
            retrieved_docs = []
            for result in results:
                doc = {
                    "id": result.get("id"),
                    "content": result.get("content"),
                    "source": result.get("source"),
                    "score": result.get("@search.score")
                }
                retrieved_docs.append(doc)
            return retrieved_docs
        except Exception as e:
            print(f"Error performing vector search: {e}")
            return []

    def hybrid_search(self, query, embedding, top_k=5):
        """Perform hybrid search (text + vector)."""
        try:
            from azure.search.documents.models import VectorizedQuery
            
            vector_query = VectorizedQuery(vector=embedding, k_nearest_neighbors=top_k, fields="embedding")
            results = self.search_client.search(
                search_text=query,
                vector_queries=[vector_query],
                top=top_k,
                include_total_count=True
            )
            
            retrieved_docs = []
            for result in results:
                doc = {
                    "id": result.get("id"),
                    "content": result.get("content"),
                    "source": result.get("source"),
                    "score": result.get("@search.score")
                }
                retrieved_docs.append(doc)
            return retrieved_docs
        except Exception as e:
            print(f"Error performing hybrid search: {e}")
            return []

    def upload_documents(self, documents):
        """Upload documents to search index."""
        try:
            self.search_client.upload_documents(documents)
            print(f"Uploaded {len(documents)} documents to index.")
            return True
        except Exception as e:
            print(f"Error uploading documents: {e}")
            return False

    def delete_documents(self, doc_ids):
        """Delete documents from index by id."""
        try:
            docs = [{"id": doc_id} for doc_id in doc_ids]
            self.search_client.delete_documents(docs)
            print(f"Deleted {len(doc_ids)} documents from index.")
            return True
        except Exception as e:
            print(f"Error deleting documents: {e}")
            return False

File: senacor-voicelive/lib/rag/vector-search.ts
/**
 * Vector Search Manager
 * Migrated from: backend/RAG/src/vector_search.py
 *
 * Azure AI Search operations for RAG vector/hybrid search.
 */

import {
  SearchClient,
  SearchIndexClient,
  KnownSearchFieldDataType,
} from '@azure/search-documents'
import { AzureKeyCredential } from '@azure/core-auth'

export interface SearchDocument {
  id: string
  content: string
  source: string
  score: number
  title?: string
}

export class VectorSearchManager {
  private searchServiceName: string
  private searchApiKey: string
  private searchIndexName: string
  private endpoint: string
  private searchClient: SearchClient<Record<string, unknown>> | null = null
  private indexClient: SearchIndexClient | null = null

  constructor() {
    this.searchServiceName = process.env.AZURE_SEARCH_SERVICE_NAME ?? ''
    this.searchApiKey = process.env.AZURE_SEARCH_API_KEY ?? ''
    this.searchIndexName = process.env.AZURE_SEARCH_INDEX_NAME ?? ''
    this.endpoint = `https://${this.searchServiceName}.search.windows.net`
  }

  /** Initialize search client and create index if it doesn't exist. */
  async initialize(): Promise<boolean> {
    try {
      const credential = new AzureKeyCredential(this.searchApiKey)

      this.searchClient = new SearchClient(
        this.endpoint,
        this.searchIndexName,
        credential,
      )

      this.indexClient = new SearchIndexClient(this.endpoint, credential)

      const indexExists = await this.indexExists()
      if (indexExists) {
        console.log(`Search index '${this.searchIndexName}' already exists.`)
      } else {
        console.log(`Search index '${this.searchIndexName}' does not exist. Creating...`)
        const created = await this.createIndex()
        if (created) {
          console.log(`Search index '${this.searchIndexName}' created successfully.`)
        } else {
          console.error(`Failed to create index '${this.searchIndexName}'.`)
          return false
        }
      }
      return true
    } catch (e) {
      console.error(`Error initializing search client: ${e}`)
      return false
    }
  }

  /** Check if index exists. */
  private async indexExists(): Promise<boolean> {
    try {
      if (!this.indexClient) return false
      await this.indexClient.getIndex(this.searchIndexName)
      return true
    } catch {
      return false
    }
  }

  /** Create search index with vector field. */
  private async createIndex(): Promise<boolean> {
    try {
      if (!this.indexClient) return false

      const startTime = performance.now()

      await this.indexClient.createIndex({
        name: this.searchIndexName,
        fields: [
          { name: 'id', type: KnownSearchFieldDataType.String, key: true },
          {
            name: 'content',
            type: KnownSearchFieldDataType.String,
            searchable: true,
            analyzerName: 'standard.lucene',
          },
          {
            name: 'source',
            type: KnownSearchFieldDataType.String,
            searchable: true,
            filterable: true,
          },
          {
            name: 'title',
            type: KnownSearchFieldDataType.String,
            searchable: true,
          },
          {
            name: 'embedding',
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            type: 'Collection(Edm.Single)' as any,
            searchable: true,
            vectorSearchDimensions: 1536,
            vectorSearchProfileName: 'myHnswProfile',
          },
        ],
        vectorSearch: {
          algorithms: [
            { name: 'myHnsw', kind: 'hnsw' },
          ],
          profiles: [
            { name: 'myHnswProfile', algorithmConfigurationName: 'myHnsw' },
          ],
        },
        semanticSearch: {
          configurations: [
            {
              name: 'default',
              prioritizedFields: {
                contentFields: [{ name: 'content' }],
              },
            },
          ],
        },
      })

      const elapsed = (performance.now() - startTime) / 1000
      console.log(`[TIMING] Search index created in ${elapsed.toFixed(2)}s`)
      return true
    } catch (e) {
      console.error(`Error creating index: ${e}`)
      return false
    }
  }

  /** Perform vector search on indexed documents. */
  async vectorSearch(embedding: number[], topK = 5): Promise<SearchDocument[]> {
    try {
      if (!this.searchClient) throw new Error('Not initialized')

      const startTime = performance.now()

      const results = await this.searchClient.search('', {
        vectorSearchOptions: {
          queries: [
            {
              kind: 'vector',
              vector: embedding,
              kNearestNeighborsCount: topK,
              fields: ['embedding'],
            },
          ],
        },
        top: topK,
        includeTotalCount: true,
      })

      const retrievedDocs: SearchDocument[] = []
      for await (const result of results.results) {
        retrievedDocs.push({
          id: String(result.document.id ?? ''),
          content: String(result.document.content ?? ''),
          source: String(result.document.source ?? ''),
          score: result.score ?? 0,
          title: result.document.title ? String(result.document.title) : undefined,
        })
      }

      const elapsed = (performance.now() - startTime) / 1000
      console.log(
        `[TIMING] Vector search completed in ${elapsed.toFixed(2)}s (retrieved ${retrievedDocs.length} documents)`,
      )
      return retrievedDocs
    } catch (e) {
      console.error(`Error performing vector search: ${e}`)
      return []
    }
  }

  /** Perform hybrid search (text + vector). */
  async hybridSearch(query: string, embedding: number[], topK = 5): Promise<SearchDocument[]> {
    try {
      if (!this.searchClient) throw new Error('Not initialized')

      const startTime = performance.now()

      const results = await this.searchClient.search(query, {
        vectorSearchOptions: {
          queries: [
            {
              kind: 'vector',
              vector: embedding,
              kNearestNeighborsCount: topK,
              fields: ['embedding'],
            },
          ],
        },
        top: topK,
        includeTotalCount: true,
      })

      const retrievedDocs: SearchDocument[] = []
      for await (const result of results.results) {
        retrievedDocs.push({
          id: String(result.document.id ?? ''),
          content: String(result.document.content ?? ''),
          source: String(result.document.source ?? ''),
          score: result.score ?? 0,
          title: result.document.title ? String(result.document.title) : undefined,
        })
      }

      const elapsed = (performance.now() - startTime) / 1000
      console.log(
        `[TIMING] Hybrid search completed in ${elapsed.toFixed(2)}s (retrieved ${retrievedDocs.length} documents)`,
      )
      return retrievedDocs
    } catch (e) {
      console.error(`Error performing hybrid search: ${e}`)
      return []
    }
  }

  /** Upload documents to search index. */
  async uploadDocuments(documents: Record<string, unknown>[]): Promise<boolean> {
    try {
      if (!this.searchClient) throw new Error('Not initialized')

      const startTime = performance.now()
      await this.searchClient.uploadDocuments(documents)
      const elapsed = (performance.now() - startTime) / 1000
      console.log(`[TIMING] Indexed ${documents.length} documents in ${elapsed.toFixed(2)}s`)
      return true
    } catch (e) {
      console.error(`Error uploading documents: ${e}`)
      return false
    }
  }

  /** Delete documents from index by id. */
  async deleteDocuments(docIds: string[]): Promise<boolean> {
    try {
      if (!this.searchClient) throw new Error('Not initialized')
      const docs = docIds.map((id) => ({ id }))
      await this.searchClient.deleteDocuments(docs)
      console.log(`Deleted ${docIds.length} documents from index.`)
      return true
    } catch (e) {
      console.error(`Error deleting documents: ${e}`)
      return false
    }
  }
}
